File: src/Plane.cpp
/*
  Copyright (C) 2011 Jon Macey

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received m_a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include "Plane.h"
#include <boost/format.hpp>
//----------------------------------------------------------------------------------------------------------------------
/// @file Plane.cpp
/// @brief implementation files for Plane class
//----------------------------------------------------------------------------------------------------------------------
namespace ngl
{

//----------------------------------------------------------------------------------------------------------------------
Plane::Plane() noexcept
{

}
//----------------------------------------------------------------------------------------------------------------------
Plane::Plane( const Vec3 &_v1, const Vec3 &_v2,  const Vec3 &_v3) noexcept
{
  setPoints(_v1,_v2,_v3);
}

//----------------------------------------------------------------------------------------------------------------------
Plane::~Plane() noexcept
{

}
//----------------------------------------------------------------------------------------------------------------------
void Plane::setPoints(const Vec3 &_v1, const Vec3 &_v2, const Vec3 &_v3) noexcept
{
	Vec3 aux1, aux2;

	aux1 = _v1 - _v2;
	aux2 = _v3 - _v2;
	m_normal = aux2.cross(aux1);
	m_normal.normalize();
	m_point=_v2;
	m_d = -(m_normal.inner(m_point));
}
//----------------------------------------------------------------------------------------------------------------------
void Plane::setNormalPoint( const Vec3 &_normal, const Vec3 &_point) noexcept
{
	m_point=_point;
	m_normal=_normal;
	m_normal.normalize();
	m_d = -(m_normal.inner(m_point));
}
//----------------------------------------------------------------------------------------------------------------------
void Plane::setFloats(Real _a,Real _b,	Real _c,	Real _d) noexcept
{
	// set the normal vector
	m_normal.set(_a,_b,_c);
	//compute the lenght of the vector
	Real l = m_normal.length();
	// normalize the vector
	m_normal.normalize();
	// and divide d by th length as well
	m_d = _d/l;
}

//----------------------------------------------------------------------------------------------------------------------
Real Plane::distance( const Vec3 &_p) const noexcept
{
	return (m_d + m_normal.inner(_p));
}

} // end of ngl namespace

File: py-ngl/src/ngl/plane.py
from .vec3 import Vec3


class Plane:
    """A mathematical plane."""

    def __init__(self, p1: Vec3 = None, p2: Vec3 = None, p3: Vec3 = None) -> None:
        self._normal = Vec3(0.0, 1.0, 0.0)
        self._point = Vec3()
        self._d = 0.0
        if p1 and p2 and p3:
            self.set_points(p1, p2, p3)

    @property
    def normal(self) -> Vec3:
        return self._normal

    @property
    def point(self) -> Vec3:
        return self._point

    @property
    def d(self) -> float:
        return self._d

    def set_points(self, p1: Vec3, p2: Vec3, p3: Vec3) -> None:
        aux1 = p1 - p2
        aux2 = p3 - p2
        self._normal = aux2.cross(aux1)
        self._normal.normalize()
        self._point = p2
        self._d = -(self._normal.inner(self._point))

    def set_normal_point(self, normal: Vec3, point: Vec3) -> None:
        self._normal = normal
        self._normal.normalize()
        self._point = point
        self._d = -(self._normal.inner(self._point))

    def set_floats(self, a: float, b: float, c: float, d: float) -> None:
        self._normal.set(a, b, c)
        length = self._normal.length()
        self._normal.normalize()
        self._d = d / length

    def distance(self, p: Vec3) -> float:
        return self._d + self._normal.inner(p)
